hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/rsql/RSQLUtility.java

```diff
@@ -594,19 +610,59 @@ private Predicate getInPredicate(final List<Object> transformedValues, final Pat
-            final List<String> outParams = new ArrayList<>();
-            for (final Object param : transformedValues) {
-                if (param instanceof String) {
-                    outParams.add(((String) param).toUpperCase());
-                }
+        private Predicate getOutPredicate(final List<Object> transformedValues, final String finalProperty,
+                final A enumField, final Path<Object> fieldPath) {
+
+            final String[] fieldNames = getSubAttributesFrom(finalProperty);
+            final List<String> outParams = transformedValues.stream().filter(String.class::isInstance)
+                    .map(String.class::cast).map(String::toUpperCase).collect(Collectors.toList());
+
+            if (isSimpleField(fieldNames, enumField.isMap())) {
+                return toNullOrNotInPredicate(fieldPath, transformedValues, outParams);
             }
-            if (!outParams.isEmpty()) {
-                return cb.not(cb.upper(pathOfString(fieldPath)).in(outParams));
+
+            clearOuterJoinsIfNotNeeded();
+
+            return toOutWithSubQueryPredicate(fieldNames, transformedValues, enumField, outParams);
+        }
+
+        private Predicate toNullOrNotInPredicate(final Path<Object> fieldPath, final List<Object> transformedValues,
+                final List<String> outParams) {
+
+            Predicate inPredicate;
+            final Path<String> pathOfString = pathOfString(fieldPath);
+
+            if (outParams.isEmpty()) {
+                inPredicate = fieldPath.in(transformedValues);
             } else {
-                return cb.not(fieldPath.in(transformedValues));
+                inPredicate = cb.upper(pathOfString).in(outParams);
+            }
 
+            return cb.or(cb.isNull(pathOfString), cb.not(inPredicate));
+        }
+
+        @SuppressWarnings({ "unchecked", "rawtypes" })
+        private Predicate toOutWithSubQueryPredicate(final String[] fieldNames, final List<Object> transformedValues,
+                final A enumField, final List<String> outParams) {
+
+            final Class<?> javaType = root.getJavaType();
+            final Subquery<?> subquery = query.subquery(javaType);
+            final Root subqueryRoot = subquery.from(javaType);
+
+            final Path innerFieldPath = getInnerFieldPath(subqueryRoot, fieldNames, enumField.isMap());
+            final Expression<String> expressionToCompare = getExpressionToCompare(innerFieldPath, enumField);
+
+            Predicate inPredicate;
+            if (!outParams.isEmpty()) {
+                inPredicate = cb.upper(expressionToCompare).in(outParams);
+            } else {
+                inPredicate = expressionToCompare.in(transformedValues);
             }
+
+            subquery.select(subqueryRoot).where(cb.and(cb.equal(root.get(enumField.identifierFieldName()),
+                    subqueryRoot.get(enumField.identifierFieldName())), inPredicate));
```

Comment:
could you please extract the equal predicate to separate method, so it is clear what the intention is